src/clima_diagnostics.jl

```diff
@@ -118,10 +119,38 @@ function DiagnosticsHandler(scheduled_diagnostics, Y, p, t; dt = nothing)
         variable = diag.variable
         isa_time_reduction = !isnothing(diag.reduction_time_func)
 
-        # The first time we call compute! we use its return value. All the subsequent times
-        # (in the callbacks), we will write the result in place. We call copy to acquire ownership
-        # of the data in case compute! returned a reference.
-        push!(storage, copy(variable.compute!(nothing, Y, p, t)))
+        # We have three cases:
+
+        # 1. The diagnostic has `compute!`. In this case, the first time we call compute! we
+        # use its return value. All the subsequent times (in the callbacks), we will write
+        # the result in place.
+        #
+        # 2a. The diagnostic has a `compute` function that returns a `Field`. In this case,
+        # we can directly copy over the result.
+        #
+        # 2b. The diagnostic has a `compute` function that returns a
+        # `Base.Broadcast.Broadcasted` (when using LazyBroadcast.jl). In this case, we have
+        # to manually materialize the result.
+
+        has_inplace_compute = !isnothing(variable.compute!)
+
+        if has_inplace_compute
+            # Case 1
+            out_field = variable.compute!(nothing, Y, p, t)
+        else
+            out_or_broadcasted = variable.compute(Y, p, t)
+            if out_or_broadcasted isa Base.Broadcast.Broadcasted
```

Comment:
```suggestion
            if out_or_broadcasted isa Base.AbstractBroadcasted
```
I think this may work more generally (e.g., in case the broadcasted is a ClimaCore stencil broadcasted)

Reply:
Added!